s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -65,6 +65,24 @@ private[alpakka] final class S3Stream(credentials: AWSCredentials,
     Source.fromFuture(request(s3Location, range).flatMap(entityForSuccess).map(_.dataBytes)).flatMapConcat(identity)
   }
 
+  def listBucket(bucket: String, prefix: Option[String] = None): Source[String, NotUsed] = {
+    def listBucketCall(continuation_token: Option[String] = None): Future[ListBucketResult] =
+      signAndGetAs[ListBucketResult](HttpRequests.listBucket(bucket, region, prefix, continuation_token))
+
+    def fileSourceFromFuture(f: Future[ListBucketResult]): Source[String, NotUsed] =
+      Source
+        .fromFuture(f)
+        .flatMapConcat((res: ListBucketResult) => {
```

Comment:
In the meantime I figured that using `Source.unfoldAsync` would probably even simpler. It allows you to thread state through between each call. So, it should basically look similar to this:

```scala
def getNextBucket(continuationToken: Option[String]): Future[Option[(String, Seq[...])]] // needs to return a future of a tuple `Some(next_continuation_token, sequence of results)` (or `None` if complete).

Source.unfoldAsync(None)(getNextBucket)
   .mapConcat(identity) // flatten sequence of results
```

Can you try that?